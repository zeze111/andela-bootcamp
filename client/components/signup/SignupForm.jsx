import React from 'react';
import PropTypes from 'prop-types';
import { Redirect } from 'react-router-dom';
import Validator from 'validatorjs';
import validations from '../../../Server/shared/validations';
import TextFieldGroup from '../common/TextFieldGroup';

class SignupForm extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      firstName: '',
      surname: '',
      email: '',
      password: '',
      password_confirmation: '',
      errors: {},
      isLoading: false
    }

    this.onChange = this.onChange.bind(this);
    this.onSubmit = this.onSubmit.bind(this);
  }

  onChange(e) {
    this.setState({ [e.target.name]: e.target.value });
  }

  isValid() {
    const validator = new Validator(this.state, validations.userRules);
    if (validator.fails()) {
      const errors = validator.errors.all()
      this.setState({ errors });
    }

    return validator.passes();
  }

  onSubmit(e) {
    e.preventDefault();


    if (this.isValid()) {
      this.setState({ errors: {}, isLoading: true });
      this.props.userSignupRequest(this.state)
        .then(() => {
          this.props.addFlashMessages({
            type: 'success',
            text: 'Your registration was successful. Welcome!'
          });
          this.setState({ redirect: true });
        })
        .catch((error) => { this.setState({ errors: error.response.data, isLoading: false }); });
    }
  }

  render() {
    const { errors } = this.state;
    const { redirect } = this.state;
    if (redirect) {
      return <Redirect to="/" />;
    }

    return (
      <div>
        <form onSubmit={this.onSubmit} className="col s5 offset-s4"> <br />
        {errors && <span className='red-text' style={{ fontSize: 16 + 'px' }}>
          {errors.message}</span>}
          <TextFieldGroup
            label="First Name"
            value={this.state.firstName}
            onChange={this.onChange}
            id="fname"
            type="text"
            name="firstName"
            error={errors.firstName}
          />
          <TextFieldGroup
            label="Surname"
            value={this.state.surname}
            onChange={this.onChange}
            id="sname"
            type="text"
            name="surname"
            error={errors.surname}
          />
          <TextFieldGroup
            label="Email"
            value={this.state.email}
            onChange={this.onChange}
            id="email"
            type="email"
            name="email"
            error={errors.email}
          />
          <TextFieldGroup
            label="Password"
            value={this.state.password}
            onChange={this.onChange}
            id="pwd"
            type="password"
            name="password"
            error={errors.password}
          />
          <TextFieldGroup
            label="Password Confirmation"
            value={this.state.password_confirmation}
            onChange={this.onChange}
            id="cf-pwd"
            type="password"
            name="password_confirmation"
            error={errors.password_confirmation}
          />
          <div className="right-align">
            <input disabled={this.state.isLoading} className="btn grey white-text" type="submit" value="Register" />
          </div> <br />
        </form>
      </div>
    );
  }
}

SignupForm.propTypes = {
  userSignupRequest: PropTypes.func.isRequired,
  addFlashMessages: PropTypes.func.isRequired
}


export default SignupForm;
