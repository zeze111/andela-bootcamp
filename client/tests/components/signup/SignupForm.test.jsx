import React from 'react';

import SignupForm from '../../../components/signup/SignupForm';
import store from '../../../store';

let props = {
  signUp: jest.fn(() => Promise.resolve())
}

const setup = () => {
  const wrapper = shallow(<SignupForm {...props} store={store} />);
  return { wrapper }
}

describe('onChange', () => {
  it('should change event', () => {
    const event = {
      target: { name: 'surname', value: 'Doe' },
    };
    const { wrapper } = setup();
    const surname = wrapper.find('#sname');
    surname.simulate('change', event);
    expect(wrapper.instance().state.surname).toBe('Doe');
  });
});

describe('onSubmit', () => {
  it('be called when form is submitted', () => {
    const { wrapper } = setup();
    const event = {
      preventDefault: jest.fn(),
    };
    wrapper.setState({
      firstName: 'Jane',
      surname: 'Doe',
      email: 'jane@yahoo.com',
      password: 'janedoe',
      password_confirmation: 'janedoe',
      errors: {},
      isLoading: false
    });
    wrapper.instance().onSubmit(event);
    expect(props.signUp.mock.calls.length).toEqual(1);
  });
});

describe('onSubmit', () => {
  it('return validation error when form is submitted', () => {
    const event = {
      preventDefault: jest.fn(),
    };
    props.signUp = jest.fn(() => Promise.reject());
    const wrapper = shallow(<SignupForm {...props} store={store} />);

    wrapper.setState({
      firstName: 'Jane',
      surname: 'Doe',
      email: 'jane@yahoo.com',
      password: 'janedoe'
    });
    wrapper.instance().onSubmit(event);
    expect(props.signUp.mock.calls.length).toEqual(0);
  });
});

describe('onSubmit', () => {
  it('return an error when form is submitted', () => {
    const event = {
      preventDefault: jest.fn(),
    };
    props.signUp = jest.fn(() => Promise.reject());
    const wrapper = shallow(<SignupForm {...props} store={store} />);

    wrapper.setState({
      firstName: 'Jane',
      surname: 'Doe',
      email: 'jane@yahoo.com',
      password: 'janedoe',
      password_confirmation: 'janedoe',
      errors: {},
      isLoading: false
    });
    wrapper.instance().onSubmit(event);
    expect(props.signUp.mock.calls.length).toEqual(1);
  });
});

describe('Test for Sign up form', () => {
  it('should render component', () => {

    const wrapper = setup();
    expect(wrapper).toMatchSnapshot()
  });
});
