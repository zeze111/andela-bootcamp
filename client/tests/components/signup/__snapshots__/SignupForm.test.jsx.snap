// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test for Sign up form should render component 1`] = `
Object {
  "wrapper": <div>
    <form
      className="col s10 m7 l6 push-s1 push-m2 push-l3"
      onSubmit={[Function]}
    >
       
      <br />
      <span
        className="red-text error-text"
      />
      <TextFieldGroup
        active=""
        error={Array []}
        id="fname"
        label="First Name"
        name="firstName"
        onChange={[Function]}
        placeholder=""
        req="*"
        type="text"
        value=""
      />
      <TextFieldGroup
        active=""
        error={Array []}
        id="sname"
        label="Surname"
        name="surname"
        onChange={[Function]}
        placeholder=""
        req="*"
        type="text"
        value=""
      />
      <TextFieldGroup
        active=""
        error={Array []}
        id="email"
        label="Email"
        name="email"
        onChange={[Function]}
        placeholder=""
        req="*"
        type="email"
        value=""
      />
      <TextFieldGroup
        active=""
        error={Array []}
        id="pwd"
        label="Password"
        name="password"
        onChange={[Function]}
        placeholder=""
        req="*"
        type="password"
        value=""
      />
      <TextFieldGroup
        active=""
        error={Array []}
        id="cf-pwd"
        label="Password Confirmation"
        name="password_confirmation"
        onChange={[Function]}
        placeholder=""
        req="*"
        type="password"
        value=""
      />
      <div
        className="right-align"
      >
        <input
          className="btn grey white-text"
          disabled={false}
          type="submit"
          value="Register"
        />
      </div>
       
      <br />
    </form>
  </div>,
}
`;
